src/halmos/cheatcodes.py

```diff
@@ -75,58 +75,92 @@ def stringified_bytes_to_bytes(hexstring: str) -> ByteVec:
+
+
+NO_PRANK = PrankResult()
+
+
+@dataclass
 class Prank:
-    addr: Any  # prank address
-    keep: bool  # start / stop prank
+    """
+    A mutable object to store the current prank context.
+
+    Because it's mutable, it must be copied across contexts.
 
-    def __init__(self, addr: Any = None, keep: bool = False) -> None:
-        if addr is not None:
-            assert_address(addr)
-        self.addr = addr
-        self.keep = keep
+    Can test for the existence of an active prank with `if prank: ...`
+
+    A prank is active if either sender or origin is set.
+
+    - prank(address) sets sender
+    - prank(address, address) sets both sender and origin
+    """
+
+    active: PrankResult = NO_PRANK  # active prank context
+    keep: bool = False  # start / stop prank
+
+    def __bool__(self) -> bool:
+        """
+        True iff either sender or origin is set.
+        """
+        return bool(self.active)
 
     def __str__(self) -> str:
-        if self.addr:
-            if self.keep:
-                return f"startPrank({str(self.addr)})"
-            else:
-                return f"prank({str(self.addr)})"
-        else:
-            return "None"
+        if not self:
+            return "no active prank"
+
+        fn_name = "startPrank" if self.keep else "prank"
+        return f"{fn_name}({str(self.active)})"
+
+    def lookup(self, to: Address) -> PrankResult:
+        """
+        If `to` is an eligible prank destination, return the active prank context.
+
+        If `keep` is False, this resets the prank context.
+        """
 
-    def lookup(self, this: Any, to: Any) -> Any:
-        assert_address(this)
         assert_address(to)
-        caller = this
         if (
-            self.addr is not None
+            self
             and not eq(to, hevm_cheat_code.address)
             and not eq(to, halmos_cheat_code.address)
         ):
-            caller = self.addr
+            result = self.active
             if not self.keep:
-                self.addr = None
-        return caller
+                self.stopPrank()
+            return result
+
+        return NO_PRANK
 
-    def prank(self, addr: Any) -> bool:
-        assert_address(addr)
-        if self.addr is not None:
+    def prank(self, sender: Address, origin: Address | None = None) -> bool:
+        assert_address(sender)
+        if self.active:
             return False
-        self.addr = addr
+
+        self.active = PrankResult(sender=sender, origin=origin)
         self.keep = False
         return True
 
-    def startPrank(self, addr: Any) -> bool:
-        assert_address(addr)
-        if self.addr is not None:
-            return False
-        self.addr = addr
-        self.keep = True
-        return True
+    def startPrank(self, sender: Address, origin: Address | None = None) -> bool:
+        result = self.prank(sender, origin)
+        self.keep = result if result else self.keep
```

Comment:
This is accurate but a bit convoluted. For better code readability, I suggest having `prank()` take an additional argument (defaulting to `False` if not provided) that is assigned to `self.keep`. Then, have `startPrank()` call `prank()` with `keep=True`.

Reply:
yeah you're right, in 17eba86